File: bookstore-ts/tests/store/book.test.ts
import { expect, test, describe } from "bun:test";
import { AuthorStore } from "../../src/store/author";
import { BookStore } from "../../src/store/book";
import { newTestDB } from "../helpers";

describe("BookStore", () => {
  test("create", () => {
    const db = newTestDB();
    const as = new AuthorStore(db);
    const bs = new BookStore(db);
    const author = as.create("Cormac McCarthy", "");
    const book = bs.create("Blood Meridian", author.id, "978-0679728757", 1985);
    expect(book.id).toBeGreaterThan(0);
    expect(book.title).toBe("Blood Meridian");
    expect(book.author_id).toBe(author.id);
  });

  describe("get", () => {
    test("exists", () => {
      const db = newTestDB();
      const as = new AuthorStore(db);
      const bs = new BookStore(db);
      const author = as.create("Haruki Murakami", "");
      const created = bs.create("Norwegian Wood", author.id, "978-0", 1987);
      const got = bs.get(created.id);
      expect(got).not.toBeNull();
      expect(got!.title).toBe("Norwegian Wood");
    });

    test("not found", () => {
      const db = newTestDB();
      const bs = new BookStore(db);
      expect(bs.get(99999)).toBeNull();
    });
  });

  describe("list", () => {
    function setup() {
      const db = newTestDB();
      const as = new AuthorStore(db);
      const bs = new BookStore(db);
      const author = as.create("Author", "");
      bs.create("Book 1", author.id, "", 2000);
      bs.create("Book 2", author.id, "", 2001);
      bs.create("Book 3", author.id, "", 2002);
      return bs;
    }

    test("all", () => {
      expect(setup().list(10, 0).length).toBe(3);
    });
    test("limit", () => {
      expect(setup().list(2, 0).length).toBe(2);
    });
    test("offset", () => {
      expect(setup().list(10, 2).length).toBe(1);
    });
  });

  test("delete", () => {
    const db = newTestDB();
    const as = new AuthorStore(db);
    const bs = new BookStore(db);
    const author = as.create("Author", "");
    const book = bs.create("To Delete", author.id, "", 2000);
    bs.delete(book.id);
    expect(bs.get(book.id)).toBeNull();
  });

  // BookSearch documents the N+1 query behaviour in search().
  // The test verifies correctness; the N+1 issue is a performance bug, not a logic bug.
  describe("search", () => {
    function setup() {
      const db = newTestDB();
      const as = new AuthorStore(db);
      const bs = new BookStore(db);
      const author = as.create("George Orwell", "");
      bs.create("1984", author.id, "", 1949);
      bs.create("Animal Farm", author.id, "", 1945);
      bs.create("Homage to Catalonia", author.id, "", 1938);
      return bs;
    }

    test("matches title", () => {
      expect(setup().search("1984").length).toBe(1);
    });
    test("partial match", () => {
      expect(setup().search("a").length).toBeGreaterThanOrEqual(2);
    });
    test("no match", () => {
      expect(setup().search("zzznomatch").length).toBe(0);
    });
  });
});

File: internal/store/book_test.go
package store_test

import (
	"testing"

	"github.com/xebia/ai-assisted-se-claude-exercise/internal/store"
)

func TestBookCreate(t *testing.T) {
	db := newTestDB(t)
	as := store.NewAuthorStore(db)
	bs := store.NewBookStore(db)

	author, _ := as.Create("Cormac McCarthy", "")
	book, err := bs.Create("Blood Meridian", author.ID, "978-0679728757", 1985)
	if err != nil {
		t.Fatalf("Create: %v", err)
	}
	if book.ID == 0 {
		t.Error("expected non-zero ID")
	}
	if book.Title != "Blood Meridian" {
		t.Errorf("Title = %q, want %q", book.Title, "Blood Meridian")
	}
	if book.AuthorID != author.ID {
		t.Errorf("AuthorID = %d, want %d", book.AuthorID, author.ID)
	}
}

func TestBookGet(t *testing.T) {
	db := newTestDB(t)
	as := store.NewAuthorStore(db)
	bs := store.NewBookStore(db)

	author, _ := as.Create("Haruki Murakami", "")
	created, _ := bs.Create("Norwegian Wood", author.ID, "978-0", 1987)

	t.Run("exists", func(t *testing.T) {
		got, err := bs.Get(created.ID)
		if err != nil {
			t.Fatalf("Get: %v", err)
		}
		if got == nil {
			t.Fatal("expected book, got nil")
		}
		if got.Title != "Norwegian Wood" {
			t.Errorf("Title = %q, want %q", got.Title, "Norwegian Wood")
		}
	})

	t.Run("not found", func(t *testing.T) {
		got, err := bs.Get(99999)
		if err != nil {
			t.Fatalf("Get: %v", err)
		}
		if got != nil {
			t.Errorf("expected nil for missing book, got %+v", got)
		}
	})
}

func TestBookList(t *testing.T) {
	db := newTestDB(t)
	as := store.NewAuthorStore(db)
	bs := store.NewBookStore(db)

	author, _ := as.Create("Author", "")
	bs.Create("Book 1", author.ID, "", 2000)
	bs.Create("Book 2", author.ID, "", 2001)
	bs.Create("Book 3", author.ID, "", 2002)

	t.Run("all", func(t *testing.T) {
		books, err := bs.List(10, 0)
		if err != nil {
			t.Fatalf("List: %v", err)
		}
		if len(books) != 3 {
			t.Errorf("len = %d, want 3", len(books))
		}
	})

	t.Run("limit", func(t *testing.T) {
		books, err := bs.List(2, 0)
		if err != nil {
			t.Fatalf("List: %v", err)
		}
		if len(books) != 2 {
			t.Errorf("len = %d, want 2", len(books))
		}
	})

	t.Run("offset", func(t *testing.T) {
		books, err := bs.List(10, 2)
		if err != nil {
			t.Fatalf("List: %v", err)
		}
		if len(books) != 1 {
			t.Errorf("len = %d, want 1", len(books))
		}
	})
}

func TestBookDelete(t *testing.T) {
	db := newTestDB(t)
	as := store.NewAuthorStore(db)
	bs := store.NewBookStore(db)

	author, _ := as.Create("Author", "")
	book, _ := bs.Create("To Delete", author.ID, "", 2000)

	if err := bs.Delete(book.ID); err != nil {
		t.Fatalf("Delete: %v", err)
	}

	got, err := bs.Get(book.ID)
	if err != nil {
		t.Fatalf("Get after delete: %v", err)
	}
	if got != nil {
		t.Error("expected nil after delete, book still exists")
	}
}

// TestBookSearch documents the N+1 query behaviour in Search().
// The test verifies correctness; the N+1 issue is a performance bug, not a logic bug.
func TestBookSearch(t *testing.T) {
	db := newTestDB(t)
	as := store.NewAuthorStore(db)
	bs := store.NewBookStore(db)

	author, _ := as.Create("George Orwell", "")
	bs.Create("1984", author.ID, "", 1949)
	bs.Create("Animal Farm", author.ID, "", 1945)
	bs.Create("Homage to Catalonia", author.ID, "", 1938)

	t.Run("matches title", func(t *testing.T) {
		results, err := bs.Search("1984")
		if err != nil {
			t.Fatalf("Search: %v", err)
		}
		if len(results) != 1 {
			t.Errorf("len = %d, want 1", len(results))
		}
	})

	t.Run("partial match", func(t *testing.T) {
		results, err := bs.Search("a")
		if err != nil {
			t.Fatalf("Search: %v", err)
		}
		if len(results) < 2 {
			t.Errorf("len = %d, want >=2", len(results))
		}
	})

	t.Run("no match", func(t *testing.T) {
		results, err := bs.Search("zzznomatch")
		if err != nil {
			t.Fatalf("Search: %v", err)
		}
		if len(results) != 0 {
			t.Errorf("len = %d, want 0", len(results))
		}
	})
}
